perf(eventstore): get_events fetches event data from nodestore

lynnagara opened the pull request:
These queries yield 10+ columns from Snuba (including all
of the tag data) and are frequently taking a long time.
Instead we now fetch the minimal column set from Snuba, then use
those IDs to populate the event bodies from nodestore. Users of
eventstore.get_events() should no longer manually bind nodes since this
has already been done.

This depends on improvements to the message / search_message field https://github.com/getsentry/sentry/pull/16473

lynnagara committed:
perf(eventstore): Fetch event data from nodestore

These queries yield 10+ columns from Snuba (including all
of the tag data) and are frequently timing out in production.
Instead we now fetch the minimal column set from Snuba, then use
those IDs to populate the event bodies from nodestore. Users of
eventstore.get_events() should no longer manually bind nodes since this
has already been done.

lynnagara commented on tests/sentry/eventstore/snuba_discover/test_backend.py:
```diff
@@ -74,45 +73,6 @@ def setUp(self):
 
         self.eventstore = SnubaDiscoverEventStorage()
 
-    def test_get_events(self):
```
These were unnecessary duplicates of snuba/test_backend.py

lynnagara commented on tests/snuba/models/test_event.py:
```diff
@@ -78,26 +77,3 @@ def test_bind_nodes(self):
         assert event.data._node_data is not None
         assert event.data["user"]["id"] == u"user1"
-
-    def test_event_with_no_body(self):
```
Since we don't have sampling as a feature anymore, I don't think this is a valid use case, and much of our product already depends on the events being present in nodestore and would otherwise be breaking if the event bodies were missing.

lynnagara committed:
fix(SnubaEvent): Fix bug in get_minimal_user

This code didn't previously work (user should be an interface of an
event), it just happens to not currently be called in any situation
where snuba data isn't populated.

lynnagara committed:
update test

lynnagara committed:
fix message

lynnagara committed:
pull in new store_event tests

fpacifici commented on src/sentry/models/event.py:
```diff
@@ -496,7 +493,9 @@ def datetime(self):
             return self.snuba_data["message"]
-        return self.data.get("message")
+        from sentry.event_manager import EventManager
+        event_manager = EventManager(self.data.data)
```
Isn't EventManager depending on Event itself ? Could we avoid this dependency by moving get_search_message ? search_message feels like a property of an event.

fpacifici commented on src/sentry/eventstore/snuba/backend.py:
```diff
@@ -62,7 +62,9 @@ def get_events(
 
         if "error" not in result:
-            return [SnubaEvent(evt) for evt in result["data"]]
+            events = [SnubaEvent(evt) for evt in result["data"]]
```
Could you elaborate on how you noticed the performance issue? Knowing that the timeout is around 30 seconds, I would be surprised if the culprit was fetching multiple columns instead and I would be less surprised if the issue was the select process, which means the queries are still likely to timeout. Could this be related to the max_thread parameter?

lynnagara replied:
I think this is more about increasing the speed of existing queries, than reducing the count of timed out queries (hopefully we don't have too many of those). I've seen a number of transactions that look similar to this this https://sentry.io/organizations/sentry/eventsv2/snuba:7c55cef95fef4fafb2444c27376596fe. The 2nd query is significantly slower than the first since it's getting all columns https://sentry.io/organizations/sentry/eventsv2/snuba:7d55be78a00d43329937c42fcba9d2d5. I've changed the get_events implementation to only ever get the minimal columns from Snuba and the rest from Nodestore.

fpacifici commented on src/sentry/eventstore/snuba/backend.py:
```diff
@@ -62,7 +62,9 @@ def get_events(
         if "error" not in result:
-            return [SnubaEvent(evt) for evt in result["data"]]
+            events = [SnubaEvent(evt) for evt in result["data"]]
+            self.bind_nodes(events)
```
This way we would be doing one request to bigtable per event. I guess, it would be more efficient to rely on nodestore.get_multi that makes one query and streams the results.
Also would the performance improvement be still relevant if we fetch like 1000 events (which means blasting bigtable with 1000 requests)

lynnagara replied:
There were 2 place where we were previously fetching more than 100 events:
- unmerge: This fetches 500 events at a time, since it was already doing bind_nodes immediately after the eventstore fetch (which has been removed) there should be no difference in performance
- nodestore deletions: Rewrote to not use eventstore since we only need to get 2 columns

fpacifici replied:
If we are doing a multi_get (which we are) I think we should be fine with that. But we should keep an eye on this.

lynnagara replied:
The only place we are adding a multi get where where we weren't (at some later point outside eventstore) is in the group deletions task, we should keep an eye on that.

lynnagara committed:
revert message change

lynnagara committed:
fix message

lynnagara committed:
remove unnecessary bind nodes

lynnagara committed:
Merge remote-tracking branch 'origin/master' into fetch-events-from-nodestore

lynnagara committed:
don't use eventstore for deletions

fpacifici commented on src/sentry/api/endpoints/group_hashes.py:
```diff
@@ -79,14 +79,17 @@ def __handle_results(self, project_id, group_id, user, results):
+                "project_id": project_id,
+            }
+        )
+        event.bind_node_data()
```
What does this become after SnubaEvent is gone ?

lynnagara replied:
This has been fixed in https://github.com/getsentry/sentry/pull/16286, will rebase to pull in those changes

fpacifici commented on src/sentry/deletions/defaults/group.py:
```diff
@@ -25,34 +26,33 @@ def chunk(self):
             referrer="deletions.group",
             orderby=["-timestamp", "-event_id"],
-        )
+        )["data"]
```
I don't think we should bypass eventstore to fetch events. See my comment on PR https://github.com/getsentry/sentry/pull/16284.

Let's first see if loading everything from eventstore is a problem. Maybe we are fine with a multi_get. If it is a problem we should revisit the issue of loading a small number of fields on a big number of events through eventstore.

lynnagara replied:
I'll revert this commit, and we can observe if it proves to be a problem before attempting something like this.

lynnagara committed:
Revert "don't use eventstore for deletions"

This reverts commit 8c12bde0c8f2d6e96761286252fcbc723daf6df9.

lynnagara committed:
Merge remote-tracking branch 'origin/master' into fetch-events-from-nodestore

lynnagara committed:
update test

lynnagara committed:
Merge remote-tracking branch 'origin/master' into fetch-events-from-nodestore

lynnagara committed:
Merge remote-tracking branch 'origin/master' into fetch-events-from-nodestore

fpacifici reviewed (changes requested):
So after the discussion we had I believe you are going to still support binding nodes in some cases. RIght ?

lynnagara committed:
Merge remote-tracking branch 'origin/master' into fetch-events-from-nodestore

lynnagara committed:
make search_message a property of event

lynnagara committed:
update deletion to use new method

lynnagara committed:
refactor

fpacifici commented on src/sentry/eventstore/base.py:
```diff
@@ -117,6 +92,11 @@ def get_events(
+    def get_unfetched_events(
+        self, filter, orderby=None, limit=100, offset=0, referrer="eventstore.get_unfetched_events"
+    ):
+        raise NotImplementedError
```
Docstring here is specifically important to explain the difference with get_events

fpacifici commented on src/sentry/eventstore/models.py:
```diff
@@ -197,6 +201,29 @@ def location(self):
             return self._snuba_data[column]
         return super(Event, self).location
 
+    def get_search_message(self, event_metadata=None, culprit=None):
```
Is this change actually needed for fetching data from nodestore ?

lynnagara replied:
It's a dependency. This was kind of a hack before since the message from snuba is different from the message from nodestore. We're no longer providing a way to explicitly say that you want the column to be loaded from snuba (a good thing since this was being abused before). I'm going to try to improve this a bit more in a separate PR.

lynnagara replied:
See https://github.com/getsentry/sentry/pull/16473 for a more proper fix for the event.message / event.search_message behaviour

fpacifici commented on src/sentry/eventstore/snuba/backend.py:
```diff
@@ -48,20 +48,59 @@ class SnubaEventStorage(EventStorage):
-        Get events from Snuba.
+        Get events from Snuba, with node data loaded.
         """
+        return self.get_events(
```
is this a call to __get_events instead ?

lynnagara replied:
uhhh, yes it should be

fpacifici reviewed (changes requested):
Failing tests are related.

lynnagara committed:
call the right method

lynnagara committed:
docstrings

lynnagara committed:
revert raw_query

lynnagara committed:
Merge remote-tracking branch 'origin/master' into fetch-events-from-nodestore

lynnagara committed:
ref: Make search_message a property of Event

Since search_message is distinct from message, we should have them as
separate properties of Event, rather than use them interchangably.

Search_message is the internal only property for search, while message is
the publicly exposed property.

We temporarily save search_message into the message field in
event_manager.save() to maintain backward compatibility with Django
event format.

lynnagara committed:
remove unnecessary import

lynnagara committed:
comment

lynnagara committed:
update docstring

lynnagara committed:
Merge remote-tracking branch 'origin/event-message-improvements' into fetch-events-from-nodestore

lynnagara committed:
move search_message to EventCommon

lynnagara committed:
Merge remote-tracking branch 'origin/event-message-improvements' into fetch-events-from-nodestore

lynnagara committed:
fix merge issue

lynnagara committed:
replace accidentally deleted function

lynnagara committed:
remove old stuff

lynnagara committed:
Merge remote-tracking branch 'origin/master' into fetch-events-from-nodestore

lynnagara committed:
temporarily try running test suite with flag flipped

lynnagara committed:
Revert "temporarily try running test suite with flag flipped"

This reverts commit 7ee79adbdcbf0f1a0cd99e1ff915894c164accd6.
